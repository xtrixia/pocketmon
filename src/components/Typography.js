import clsx from "clsx";
import { css } from "@emotion/css";

import { SPACINGS } from "../root/spacings";
import { COLORS } from "../root/colors";

function Typography({ className, variant, ...props }) {
  let Component = generateTag(variant);

  return (
    <Component
      className={clsx(
        className,
        css`
          ${generateStyle(variant)}
          overflow-wrap: anywhere;
        `
      )}
      {...props}
    />
  );
}

function generateStyle(variant) {
  switch (variant) {
    case "h1":
      return {
        fontWeight: 300,
        fontSize: "6rem",
        lineHeight: 1.167,
        letterSpacing: "-0.01562em",
      };
    case "h2":
      return {
        fontWeight: 300,
        fontSize: "3.75rem",
        lineHeight: 1.2,
        letterSpacing: "-0.00833em",
      };
    case "h3":
      return {
        fontWeight: 300,
        fontSize: "3rem",
        lineHeight: 1.167,
        letterSpacing: 0,
      };
    case "h4":
      return {
        fontWeight: 400,
        fontSize: "2.125rem",
        lineHeight: 1.235,
        letterSpacing: "0.00735",
      };
    case "h5":
      return {
        fontWeight: 400,
        fontSize: "1.5rem",
        lineHeight: 1.334,
        letterSpacing: 0,
      };
    case "h6":
      return {
        fontWeight: 500,
        fontSize: "1.25rem",
        lineHeight: 1.6,
        letterSpacing: "0.0075em",
      };
    case "subtitle":
      return {
        fontWeight: 400,
        fontFamily: "Karla, sans-serif",
        fontSize: "1rem",
        lineHeight: 1.75,
        letterSpacing: "0.00938em",
      };
    case "subheading1":
      return {
        padding: `${SPACINGS.md} 0`,
        fontFamily: `"Lora", serif`,
        fontWeight: 300,
        fontSize: "3.75rem",
        lineHeight: 1.2,
        letterSpacing: "-0.00833em",
      };
    case "subheading2":
      return {
        padding: `${SPACINGS.md} 0`,
        fontFamily: `"Lora", serif`,
        color: COLORS.dark,
        fontWeight: 400,
        fontSize: "2rem",
        lineHeight: 1.235,
        letterSpacing: "0.00735",
      };
    default:
      return {
        fontWeight: 400,
        fontSize: "1rem",
        lineHeight: 1.5,
        letterSpacing: "0.00938em",
      };
  }
}

function generateTag(variant) {
  const variants = ["h1", "h2", "h3", "h4", "h5", "h6"];
  const subheadings = ["subheading1", "subheading2"];

  switch (variant) {
    case variants.includes(variant):
      return variant;
    case subheadings.includes(variant):
      return variant === "subheading1" ? "h2" : "h4";
    default:
      return "p";
  }
}

export default Typography;
