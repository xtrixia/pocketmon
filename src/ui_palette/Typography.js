import clsx from "clsx";
import { css } from "@emotion/css";

function Typography({ className, variant, ...props }) {
  let Component = generateTag(variant);

  return (
    <Component
      className={clsx(
        className,
        css`
          ${generateStyle(variant)}
        `
      )}
      {...props}
    />
  );
}

function generateStyle(variant) {
  switch (variant) {
    case "h1":
      return {
        fontWeight: 300,
        fontSize: "6rem",
        lineHeight: 1.167,
        letterSpacing: "-0.01562em",
      };
    case "h2":
      return {
        fontWeight: 300,
        fontSize: "3.75rem",
        lineHeight: 1.2,
        letterSpacing: "-0.00833em",
      };
    case "h3":
      return {
        fontWeight: 300,
        fontSize: "3rem",
        lineHeight: 1.167,
        letterSpacing: 0,
      };
    case "h4":
      return {
        fontWeight: 400,
        fontSize: "2.125rem",
        lineHeight: 1.235,
        letterSpacing: "0.00735",
      };
    case "h5":
      return {
        fontWeight: 400,
        fontSize: "1.5rem",
        lineHeight: 1.334,
        letterSpacing: 0,
      };
    case "h6":
      return {
        fontWeight: 500,
        fontSize: "1.25rem",
        lineHeight: 1.6,
        letterSpacing: "0.0075em",
      };
    case "subtitle":
      return {
        fontWeight: 400,
        fontSize: "1rem",
        lineHeight: 1.75,
        letterSpacing: "0.00938em",
      };
    default:
      return {
        fontWeight: 400,
        fontSize: "1rem",
        lineHeight: 1.5,
        letterSpacing: "0.00938em",
      };
  }
}

function generateTag(variant) {
  const variants = ["h1", "h2", "h3", "h4", "h5", "h6"];
  return variants.includes(variant) ? variant : "p";
}

export default Typography;
