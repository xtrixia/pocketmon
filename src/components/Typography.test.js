import React from "react";
import renderer from "react-test-renderer";

import Typography from "./Typography";

test("Typography should render all variants", () => {
  const variantMapping = [
    "h1",
    "h2",
    "h3",
    "h4",
    "h5",
    "h6",
    "subheading1",
    "subheading2",
    "subtitle",
    "body",
  ];

  variantMapping.forEach((variant) => {
    const component = renderer.create(
      <Typography variant={variant}>Hello</Typography>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    // expect(tree).toHaveProperty("type", variant);
  });
});

test("Typography render text passed from props", () => {
  const component = renderer.create(<Typography>Hello world</Typography>);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree.children[0]).toEqual("Hello world");
});

test("Typography should render variant body by default", () => {
  const component = renderer.create(<Typography>Hello world</Typography>);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  // we did an abstraction for variant body using <p>
  expect(tree.type).toEqual("p");
});
