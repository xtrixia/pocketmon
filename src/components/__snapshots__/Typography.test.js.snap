// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography render text passed from props 1`] = `
<p
  className="css-qgsv95"
>
  Hello world
</p>
`;

exports[`Typography should render all variants 1`] = `
<p
  className="css-udnpao"
>
  Hello
</p>
`;

exports[`Typography should render all variants 2`] = `
<p
  className="css-1alkoj4"
>
  Hello
</p>
`;

exports[`Typography should render all variants 3`] = `
<p
  className="css-1hr3cp7"
>
  Hello
</p>
`;

exports[`Typography should render all variants 4`] = `
<p
  className="css-1i2j11s"
>
  Hello
</p>
`;

exports[`Typography should render all variants 5`] = `
<p
  className="css-1ke7tdd"
>
  Hello
</p>
`;

exports[`Typography should render all variants 6`] = `
<p
  className="css-3wutzz"
>
  Hello
</p>
`;

exports[`Typography should render all variants 7`] = `
<p
  className="css-cymwmk"
>
  Hello
</p>
`;

exports[`Typography should render all variants 8`] = `
<p
  className="css-ebjej3"
>
  Hello
</p>
`;

exports[`Typography should render all variants 9`] = `
<p
  className="css-y31snc"
>
  Hello
</p>
`;

exports[`Typography should render all variants 10`] = `
<p
  className="css-qgsv95"
>
  Hello
</p>
`;

exports[`Typography should render variant body by default 1`] = `
<p
  className="css-qgsv95"
>
  Hello world
</p>
`;
